protocol/sip/helper.go

```diff
@@ -0,0 +1,509 @@
+		if err != nil {
+			output.PrintFrameworkError("Parsing response %+v with body: %s", msg, err.Error())
+
+			return nil, false
+		}
+	}
+
+	return msg, true
+}
+
+// Sends a UDP message and returns the response.
+//
+// If 'amount' is set to 0, it will use the recommended size for UDP messages (1300 bytes).
+func SendAndReceiveUDP(
+	conn *net.UDPConn, req sip.Message,
+) (*sip.Response, bool) {
+	if conn == nil {
+		output.PrintfFrameworkError(errMsgRequiredParam, "conn")
+
+		return nil, false
+	}
+	if req == nil {
+		output.PrintfFrameworkError(errMsgRequiredParam, "req")
+
+		return nil, false
+	}
+	ok := protocol.UDPWrite(conn, []byte(req.String()))
+	if !ok {
+		output.PrintfFrameworkError("Writing message %s to the socket", req.String())
+
+		return nil, false
+	}
+	// To discard requests coming from the server, like OPTIONS or NOTIFY.
+	for {
+		respMsg, ok := ReadMessageUDP(conn)
+		if !ok {
+			output.PrintFrameworkError("Reading response from the socket")
+
+			return nil, false
+		}
+		resp, ok := respMsg.(*sip.Response)
+		if !ok {
+			output.PrintfFrameworkDebug("Response is not a valid: %+v", respMsg)
+
+			continue
+		}
+
+		return resp, true
+	}
+}
+
+// Returns a SIP message from a UDP connection using the message length
+// defined in the RFC 3261.
+// https://datatracker.ietf.org/doc/html/rfc3261#section-18.1.1
+func ReadMessageUDP(conn *net.UDPConn) (sip.Message, bool) {
+	resp := make([]byte, UDPMessageLength)
+	count, err := conn.Read(resp)
+	if err != nil {
+		output.PrintFrameworkError("Failed to read from the socket: " + err.Error())
+
+		return nil, false
+	}
+	resp = resp[:count]
+	msg, err := sip.ParseMessage(resp)
+	if err != nil {
+		output.PrintfFrameworkError("Parsing response %+v: %s", resp, err.Error())
+
+		return nil, false
+	}
+
+	return msg, true
+}
+
+// Returns a generic well-formed request. Useful to start the communication.
+//
+// Depending on the method (if not set in 'opts') required headers and body content
+// is automatically added.
+func NewSipRequest(
+	method sip.RequestMethod, host string, opts *NewSipRequestOpts,
+) (*sip.Request, bool) {
+	if method == "" {
+		output.PrintfFrameworkError(errMsgRequiredParam, "method")
+
+		return nil, false
+	}
+	if host == "" {
+		output.PrintfFrameworkError(errMsgRequiredParam, "host")
+
+		return nil, false
+	}
+	if opts == nil {
+		opts = &NewSipRequestOpts{}
+	}
+	lHost := host
+	if opts.LocalHost != "" {
+		lHost = opts.LocalHost
+	}
+	lPort := opts.Port
+	if opts.LocalPort != 0 {
+		lPort = opts.LocalPort
+	}
+	toURI := sip.Uri{
+		User: opts.ToUser,
+		Host: host,
+		Port: opts.Port,
+	}
+	fromURI := sip.Uri{
+		User:     opts.User,
+		Password: opts.Password,
+		Host:     lHost,
+		Port:     lPort,
+	}
+	req := sip.NewRequest(method, toURI)
+	if opts.Transport != 0 {
+		req.SetTransport(opts.Transport.String())
+	}
+	callID, err := uuid.NewRandom()
+	if err != nil {
+		output.PrintfFrameworkError("Generating UUID: %s", err.Error())
```

Comment:
This might be a bit of a nitpick, but the error messages don't always seem to indicate that something is wrong. I _personally_ would use something like `Could not generate UUID: %s` only because I know some of the propagated error messages can be unclear. Since it's on the framework error it *should* be clear enough so if anyone else wants to hit approve and ignore that I'm totally fine with it.